packages/dom-expressions/src/jsx.d.ts

```diff
@@ -57,13 +57,26 @@ export namespace JSX {
     $ServerOnly?: boolean;
   }
   interface Directives {}
+  interface DirectiveFunctions {}
   interface ExplicitProperties {}
   interface ExplicitAttributes {}
   interface CustomEvents {}
   interface CustomCaptureEvents {}
   type DirectiveAttributes = {
     [Key in keyof Directives as `use:${Key}`]?: Directives[Key];
   };
+  type DirectiveFunctionAttributes<T> = {
+    [Key in keyof DirectiveFunctions as `use:${Key}`]?:
+      // If function has no arguments, value is arbitrary
+      Parameters<DirectiveFunctions[Key]>['length'] extends 0 ? any :
+      // Otherwise (at least one argument), check its type matches element type
+      T extends Parameters<DirectiveFunctions[Key]>[0] ? (
+        // If function has only one argument, value is arbitrary
+        Parameters<DirectiveFunctions[Key]>['length'] extends 1 ? any :
+        // Otherwise (at least two arguments), value must match second argument
+        Parameters<DirectiveFunctions[Key]>[1]
+      ) : never;
+  };
```

Comment:
I think this isn't correct; `Parameters<DirectiveFunctions[Key]>[1]` will give `Accessor<Signal<string>>` for the example provided, instead of simply `Signal<string>` as it should. 

I suggest this definition instead, which uses `infer` directly instead of using `Parameters`:

```ts
    type DirectiveFunctionAttributes<T> = {
      [K in keyof DirectiveFunctions as `use:${K}`]?: DirectiveFunctions[K] extends (
        el: infer E, // E will be unknown if not provided
        accessor: infer A // A will be unknown if not provided
      ) => void
        ? T extends E
          ? A extends Accessor<infer V>
            ? V // if it's an Accessor, the type should be the type the Accessor wraps
            : true // if it isn't, assume the second parameter isn't provided
          : never
        : never;
    };
```

`true` corresponds to `use:directive` with no parameters provided.

Users not typing the directive function with `Accessor` will simply see `type is not assignable to 'true'` which isn't very informative, but there's no way to throw a custom error here other than with a string literal (so instead of going `true` immediately, do `unknown extends A ? true : "directive parameters must be wrapped in an Accessor (() => your type here)"`)